### PR TITLE
MixStyle Tandem Feature Regularization for OOD Generalization

### PR DESCRIPTION
**Pull request by tcapelle**
## Hypothesis

The model's tandem OOD performance (p_tan) is bottlenecked by **overfitting to the feature statistics of training-distribution tandem shapes**. When NACA6416 appears in tandem at test time, its intermediate features have different mean/variance from training tandem features, causing degraded predictions.

**MixStyle** (Zhou et al., ICLR 2021) addresses this by mixing feature statistics (mean + std) between different samples during training, creating "virtual" samples with novel statistical profiles. Applied selectively to tandem samples, this forces the model to handle a broader range of tandem feature distributions — including distributions closer to what NACA6416 tandem produces at test time.

**Why this is NOT the same as SWD (#2175):** SWD tried to ALIGN tandem and single-foil distributions (forcing them to be the same). That destroyed the tandem-specific signal. MixStyle does the opposite — it creates DIVERSITY within the tandem domain by interpolating feature statistics between different tandem samples. The tandem signal is preserved; only the statistical "style" changes.

**Key difference from input-level augmentation** (aug_dsdf2, aug_gap_stagger): Those augmentations perturb the INPUT space. MixStyle operates in the FEATURE space after the Transolver blocks have processed the input. Feature-space regularization affects how the model generalizes to ALL unseen inputs (including far-OOD), not just inputs near the training distribution.

**Reference:** Zhou et al. "Domain Generalization with MixStyle" (ICLR 2021, arXiv:2104.02008)

### Step 1 — Add CLI flags

```python
parser.add_argument('--mixstyle', action='store_true', help='Enable MixStyle feature regularization')
parser.add_argument('--mixstyle_alpha', type=float, default=0.3, help='Beta distribution alpha for MixStyle mixing coefficient')
parser.add_argument('--mixstyle_prob', type=float, default=0.5, help='Probability of applying MixStyle per forward pass')
```

### Step 2 — Add MixStyle logic in `Transolver.forward()`

Insert the following **after the non-last blocks loop** (after `for block in self.blocks[:-1]: ...`, around line ~902) and **before** `fx_deep = fx`:

```python
# MixStyle: mix feature statistics between tandem samples for OOD regularization
if self.training and getattr(self, '_mixstyle', False) and is_tandem is not None:
    import random as _rand
    if _rand.random() < self._mixstyle_prob:
        tandem_mask_flat = (is_tandem.view(-1) > 0.5)  # [B]
        n_tandem = tandem_mask_flat.sum().item()
        if n_tandem >= 2:
            tandem_idx = tandem_mask_flat.nonzero(as_tuple=True)[0]
            perm = tandem_idx[torch.randperm(n_tandem, device=fx.device)]

            eps = 1e-6
            # Per-sample statistics over spatial dimension
            mu = fx[tandem_idx].mean(dim=1, keepdim=True)          # [n_tandem, 1, H]
            sigma = fx[tandem_idx].std(dim=1, keepdim=True) + eps  # [n_tandem, 1, H]
            mu_perm = fx[perm].mean(dim=1, keepdim=True)
            sigma_perm = fx[perm].std(dim=1, keepdim=True) + eps

            lam = torch.distributions.Beta(
                self._mixstyle_alpha, self._mixstyle_alpha
            ).sample().item()

            # Normalize → mix statistics → denormalize
            fx_normed = (fx[tandem_idx] - mu) / sigma
            mu_mix = lam * mu + (1 - lam) * mu_perm
            sigma_mix = lam * sigma + (1 - lam) * sigma_perm
            fx = fx.clone()  # avoid in-place modification issues with autograd
            fx[tandem_idx] = fx_normed * sigma_mix + mu_mix
```

**Also set the attributes on the model after construction (in the training setup):**
```python
if args.mixstyle:
    model._mixstyle = True
    model._mixstyle_alpha = args.mixstyle_alpha
    model._mixstyle_prob = args.mixstyle_prob
else:
    model._mixstyle = False
```

### Step 3 — Log MixStyle activation

Every 20 steps, log whether MixStyle was applied (for debugging):
```python
if args.mixstyle and step % 20 == 0:
    wandb.log({"mixstyle/applied": 1 if mixstyle_was_applied else 0}, step=global_step)
```

### Step 4 — Run 2 configs, 2 seeds each

Use `--wandb_group askeladd-mixstyle`.

**Config A — moderate mixing (alpha=0.3, prob=0.5):**
```bash
cd cfd_tandemfoil && python train.py --agent askeladd \
  --wandb_name "askeladd/mixstyle-a03-p05" --wandb_group askeladd-mixstyle \
  --mixstyle --mixstyle_alpha 0.3 --mixstyle_prob 0.5 \
  --asinh_pressure --asinh_scale 0.75 --field_decoder --adaln_output --use_lion --lr 2e-4 \
  --aug aoa_perturb --aug_full_dsdf_rot --high_p_clamp --n_layers 3 --slice_num 96 \
  --tandem_ramp --domain_layernorm --domain_velhead --ema_decay 0.999 --weight_decay 5e-5 \
  --cosine_T_max 160 --pressure_first --pressure_deep \
  --residual_prediction --surface_refine --surface_refine_hidden 192 --surface_refine_layers 3 \
  --aft_foil_srf --aug_gap_stagger_sigma 0.02 --aug_dsdf2_sigma 0.05 \
  --pcgrad_3way --pcgrad_extreme_pct 0.15 --gap_stagger_spatial_bias \
  --seed 42
```
(Repeat with `--seed 73`)

**Config B — stronger mixing (alpha=0.5, prob=1.0):**
```bash
cd cfd_tandemfoil && python train.py --agent askeladd \
  --wandb_name "askeladd/mixstyle-a05-p10" --wandb_group askeladd-mixstyle \
  --mixstyle --mixstyle_alpha 0.5 --mixstyle_prob 1.0 \
  --asinh_pressure --asinh_scale 0.75 --field_decoder --adaln_output --use_lion --lr 2e-4 \
  --aug aoa_perturb --aug_full_dsdf_rot --high_p_clamp --n_layers 3 --slice_num 96 \
  --tandem_ramp --domain_layernorm --domain_velhead --ema_decay 0.999 --weight_decay 5e-5 \
  --cosine_T_max 160 --pressure_first --pressure_deep \
  --residual_prediction --surface_refine --surface_refine_hidden 192 --surface_refine_layers 3 \
  --aft_foil_srf --aug_gap_stagger_sigma 0.02 --aug_dsdf2_sigma 0.05 \
  --pcgrad_3way --pcgrad_extreme_pct 0.15 --gap_stagger_spatial_bias \
  --seed 42
```
(Repeat with `--seed 73`)

Run all 4 in parallel. Report p_in, p_oodc, p_tan, p_re per config (2-seed avg).

### Implementation notes

- **Only mix tandem samples** — single-foil samples must NOT be touched by MixStyle. Use `is_tandem` to identify tandem samples in the batch.
- **Only during training** — MixStyle is disabled at eval/test time (`self.training` guard).
- **No label mixing** — MixStyle only changes feature statistics. Targets/labels remain unchanged. This is the key difference from Manifold Mixup.
- **The `fx.clone()` is essential** — without it, in-place modification of `fx[tandem_idx]` will break autograd.
- **If `torch.compile` issues arise**: wrap the MixStyle block in `if not torch.compiler.is_compiling():` as a fallback.
- **Expected VRAM**: identical to baseline (~43 GB). MixStyle adds negligible compute.

### Diagnostics

- If p_tan gets WORSE with Config B but not Config A → alpha=0.5 too aggressive, try alpha=0.1.
- If BOTH configs show no change → feature statistics already well-regularized by existing augmentation. Clean negative.
- Log `lam` values in a few batches to verify Beta distribution sampling.

## Baseline

| Metric | Baseline (2-seed avg) |
|--------|----------------------|
| p_in | 13.05 |
| p_oodc | 7.70 |
| **p_tan** | **28.60** |
| p_re | 6.55 |

Baseline W&B runs: d7l91p0x (seed 42, p_tan=28.9), j9btfx09 (seed 73, p_tan=28.3)

**Reproduce baseline:**
```bash
cd cfd_tandemfoil && python train.py --agent <name> --wandb_name "<name>/baseline-gsb-pcgrad" \
  --asinh_pressure --asinh_scale 0.75 --field_decoder --adaln_output --use_lion --lr 2e-4 \
  --aug aoa_perturb --aug_full_dsdf_rot --high_p_clamp --n_layers 3 --slice_num 96 \
  --tandem_ramp --domain_layernorm --domain_velhead --ema_decay 0.999 --weight_decay 5e-5 \
  --cosine_T_max 160 --pressure_first --pressure_deep \
  --residual_prediction --surface_refine --surface_refine_hidden 192 --surface_refine_layers 3 \
  --aft_foil_srf --aug_gap_stagger_sigma 0.02 --aug_dsdf2_sigma 0.05 \
  --pcgrad_3way --pcgrad_extreme_pct 0.15 --gap_stagger_spatial_bias
```

### CODE REVIEW

**Comment by tcapelle** [comment]
STUDENT askeladd:

## Results

### Summary

MixStyle feature regularization **significantly degrades** all metrics, especially tandem transfer. The hypothesis that mixing tandem feature statistics would improve OOD generalization is not supported — both configs make things worse.

### Individual Run Metrics (Surface MAE p, denormalized)

| Config | Seed | W&B Run | p_in | p_tan | p_oodc | p_re | Best Epoch |
|--------|------|---------|------|-------|--------|------|------------|
| A (α=0.3, p=0.5) | 42 | `s79jy7si` | 13.6 | 33.3 | 8.6 | 7.1 | 146 |
| A (α=0.3, p=0.5) | 73 | `6d382ygc` | 13.4 | 34.4 | 8.4 | 6.8 | 146 |
| B (α=0.5, p=1.0) | 42 | `3prht0km` | 17.3 | 36.0 | 11.0 | 8.5 | 140 |
| B (α=0.5, p=1.0) | 73 | `ov3k8rfp` | 14.8 | 36.1 | 9.6 | 7.3 | 141 |

### 2-Seed Averages vs Baseline

| Config | p_in | p_tan | p_oodc | p_re |
|--------|------|-------|--------|------|
| **Baseline** | **13.05** | **28.60** | **7.70** | **6.55** |
| MixStyle A (α=0.3, p=0.5) | 13.50 (+3.4%) | 33.85 (+18.4%) | 8.50 (+10.4%) | 6.95 (+6.1%) |
| MixStyle B (α=0.5, p=1.0) | 16.05 (+23.0%) | 36.05 (+26.0%) | 10.30 (+33.8%) | 7.90 (+20.6%) |

### Train Commands

```bash
# Config A: alpha=0.3, prob=0.5
CUDA_VISIBLE_DEVICES=0 python train.py --agent askeladd --wandb_name "askeladd/mixstyle-a03-p05" --wandb_group askeladd-mixstyle \
  --mixstyle --mixstyle_alpha 0.3 --mixstyle_prob 0.5 \
  --asinh_pressure --asinh_scale 0.75 --field_decoder --adaln_output --use_lion --lr 2e-4 \
  --aug aoa_perturb --aug_full_dsdf_rot --high_p_clamp --n_layers 3 --slice_num 96 \
  --tandem_ramp --domain_layernorm --domain_velhead --ema_decay 0.999 --weight_decay 5e-5 \
  --cosine_T_max 160 --pressure_first --pressure_deep \
  --residual_prediction --surface_refine --surface_refine_hidden 192 --surface_refine_layers 3 \
  --aft_foil_srf --aug_gap_stagger_sigma 0.02 --aug_dsdf2_sigma 0.05 \
  --pcgrad_3way --pcgrad_extreme_pct 0.15 --gap_stagger_spatial_bias --seed 42

# Config B: alpha=0.5, prob=1.0 (same baseline flags, different mixstyle params)
```

### Peak Memory

~43 GB VRAM per run (identical to baseline)

### W&B Run IDs

- `s79jy7si` (A, seed 42), `6d382ygc` (A, seed 73)
- `3prht0km` (B, seed 42), `ov3k8rfp` (B, seed 73)
- W&B group: `askeladd-mixstyle`

### What happened

**The hypothesis clearly did not work.** MixStyle significantly degrades all metrics across both configs:

1. **Config A (moderate, α=0.3, p=0.5)** increases p_tan by 18.4% (33.85 vs 28.60 baseline). This is a substantial regression, well beyond seed variance.

2. **Config B (aggressive, α=0.5, p=1.0)** is even worse — p_tan +26.0%, p_oodc +33.8%, p_in +23.0%. The damage scales with mixing strength.

3. **Why it failed:** MixStyle operates by mixing feature-level statistics (mean + std) between different tandem samples. But in this CFD context, the feature statistics encode critical physics information — things like pressure magnitude scales, velocity regimes, and geometric conditioning that are fundamentally different between tandem configurations. Mixing these statistics creates "impossible" feature distributions that don't correspond to any physical configuration, degrading the model's ability to learn the underlying physics.

4. **Key difference from vision domain generalization:** In image classification, "style" (texture, color statistics) is often a nuisance variable that should be ignored for robust classification. In CFD, the "style" (field magnitude statistics) IS the physics — it directly encodes the flow state. MixStyle's core assumption (that feature statistics carry domain-specific nuisance information) is violated here.

5. **Implementation note:** Statistics were properly detached per the reference implementation (Zhou et al. 2021, official Dassl codebase). The `@torch.compiler.disable` decorator was needed to bypass `torch.compile` graph issues with the dynamic operations (Beta sampling, nonzero indexing, randperm).

### Note on verification NaN bug

Same pre-existing verification NaN issue as observed in PR #2175 — not related to MixStyle changes.

### Suggested follow-ups

1. **Physics-aware mixing:** Instead of mixing ALL feature statistics, only mix statistics of "style-like" features (e.g., the scale/shift parameters of domain_layernorm) while preserving "content-like" features (geometric embeddings, pressure-related channels). This requires identifying which hidden dimensions encode style vs. content.

2. **Adversarial feature augmentation** (AFA): Rather than mixing statistics between real samples, add adversarial perturbations to feature statistics along the direction that maximizes loss. This creates hard examples that improve robustness without creating physically impossible feature distributions.

3. **Dropout as implicit style augmentation:** The existing model may already get sufficient feature-space regularization from dropout + domain_layernorm + the input-level augmentations. The explicit feature-space mixing from MixStyle may be redundant or counterproductive. Consider testing whether removing domain_layernorm or reducing augmentation strength worsens p_tan to understand the contribution of existing regularization.

4. **This result combined with SWD (#2175)** strongly suggests that the tandem feature representations are physically meaningful and should NOT be perturbed — alignment (SWD) or augmentation (MixStyle) in feature space both hurt. Future tandem transfer work should focus on the output/decoder side rather than intermediate features.

**Comment by tcapelle** [comment]
ADVISOR: Closing PR #2188 because ADVISOR: Closing — clear dead end. Both configs significantly degraded all metrics: Config A (α=0.3) p_tan +18.4%, Config B (α=0.5) p_tan +26.0%, with damage scaling with mixing strength.

Your analysis nails the core issue: in CFD, feature statistics (mean, std) encode the physical flow state — pressure magnitudes, velocity regimes, geometric conditioning. MixStyle's assumption that feature statistics carry domain-specific 'nuisance information' is violated here. This is fundamentally different from the vision domain generalization setting.

Combined with SWD (#2175) and DSDF TTA (#2189), we now have strong evidence: **three consecutive feature-distribution-manipulation experiments all failed catastrophically.** The tandem feature representations are physically meaningful and must NOT be perturbed at any level (input, intermediate features, or statistics). Future OOD tandem work should focus on the output/decoder side or on providing the model with genuinely new information about tandem geometry.

Your suggested follow-up #4 is spot-on and will inform our next experiments. Great work on the thorough analysis. Moving you to a new experiment..
